### PR TITLE
LibJS: Validation of NanosecondsToDays and getOffsetNanoseconds

### PR DESCRIPTION
**Pull request by moustafaraafat**
Ticks off two boxes in https://github.com/SerenityOS/serenity/issues/15525

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp [hunk 1]
```diff
@@ -538,7 +538,23 @@ ThrowCompletionOr<NanosecondsToDaysResult> nanoseconds_to_days(VM& vm, Crypto::S
...
+    if (days > 0 && sign == -1)
+        return vm.throw_completion<RangeError>(ErrorType::TemporalInvalidEpochNanoseconds);
+
+    // 21. If nanoseconds < 0 and sign = 1, throw a RangeError exception.
+    if (nanoseconds.is_negative() && sign == 1)
+        return vm.throw_completion<RangeError>(ErrorType::TemporalInvalidEpochNanoseconds);
+
+    // 22. If nanoseconds > 0 and sign = -1, throw a RangeError exception.
+    if (!nanoseconds.is_negative() && sign == -1)
+        return vm.throw_completion<RangeError>(ErrorType::TemporalInvalidEpochNanoseconds);
```
Comment (davidot) [inline Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp:555]: This won't handle `nanoseconds = 0` correctly.

Also just for clarity you might want to add an explicit check for `0` in the if above (step 21)

File: Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp [hunk 2]
```diff
@@ -554,7 +554,12 @@ ThrowCompletionOr<NanosecondsToDaysResult> nanoseconds_to_days(VM& vm, Crypto::S
     if (!nanoseconds.is_negative() && sign == -1)
         return vm.throw_completion<RangeError>(ErrorType::TemporalInvalidEpochNanoseconds);
 
-    // 23. Return the Record { [[Days]]: days, [[Nanoseconds]]: nanoseconds, [[DayLength]]: abs(dayLengthNs) }.
+    // 23. If abs(nanoseconds) ≥ abs(dayLengthNs), throw a RangeError exception.
+    auto compare_result = nanoseconds.compare_to_double(fabs(day_length_ns.to_double()));
```
Comment (davidot) [inline Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp:558]: Unless I'm missing something here you don't do the abs part of `abs(nanoseconds)`.

**Review by davidot — CHANGES_REQUESTED** [review]
Found two issues

File: Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp (hunk 2 above)
Comment (davidot) [reply Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp:558]: I see that `UnsignedBigInteger` doesn't have a `compare_to_double` function, so just matching the sign of `nanoseconds might be easier? Or you could temporarily negate nanoseconds

Comment (moustafaraafat) [reply Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp:558]: I will go with the negation, and maybe in the future will add compare_to_double() to UnsignedBigInteger.

File: Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp (hunk 1 above)
Comment (moustafaraafat) [reply Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp:555]: Sorry, I didn't get your second point.

**Review by davidot — APPROVED** [review]
LGTM

File: Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp [hunk 3]
```diff
@@ -538,7 +538,29 @@ ThrowCompletionOr<NanosecondsToDaysResult> nanoseconds_to_days(VM& vm, Crypto::S
         }
     }
 
-    // 19. Return the Record { [[Days]]: days, [[Nanoseconds]]: nanoseconds, [[DayLength]]: abs(dayLengthNs) }.
+    // 19. If days < 0 and sign = 1, throw a RangeError exception.
+    if (days < 0 && sign == 1)
+        return vm.throw_completion<RangeError>(ErrorType::TemporalInvalidEpochNanoseconds);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp:543]: This error type makes no sense in this case as the nanoseconds might have come from a `Duration`, which has nothing to do with the epoch. Same below ofc.

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp (hunk 3 above)
Comment (moustafaraafat) [reply Userland/Libraries/LibJS/Runtime/Temporal/ZonedDateTime.cpp]: Sorry, that was wrong. Done now, but maybe it's too verbose, and we can combine the 3 into something like "TemporalNanosecondsToDaysConversionError".

**Review by linusg — APPROVED** [review]
Thanks, LGTM!

**Pull request merged**